lib/puppet/type/hocon_setting.rb

```diff
@@ -22,8 +22,36 @@
     end
   end
 
-  newproperty(:value) do
+  newproperty(:type) do
+    desc "The value type"
+    # This property has no default. If it is not supplied, the validation of the "value"
+    # property will set one automatically.
+  end
+
+  newproperty(:value, :array_matching => :all) do
     desc 'The value of the setting to be defined.'
+
+    validate do |val|
+      case @resource[:type]
+        when nil
+          if @shouldorig.is_a?(Array) and @shouldorig.size > 1
+            @resource[:type] = 'array'
+          else
+            case val.class.to_s.downcase.to_sym
+              when :trueclass, :falseclass
+                @resource[:type] = 'boolean'
+              else
+                @resource[:type] = 'string'
+            end
+          end
+        when 'hash', 'array', 'text'
+          # we're just leaving these values alone
+        else
+          if @shouldorig.is_a?(Array) and @shouldorig.size > 1
+            raise "Array provided, but type specified as #{@resource[:type]}"
+          end
+      end
```

Comment:
i was reading it a couple of times looking for a way to simplify it, or collapse the paths, but there's too much going on here in order to be able to do that :\